File: scala/first.scala
object Hello{
	def main(args: Array[String]): Unit ={	
		//Print "hello, world"	
		println("Hello, world!")
		
		//Define variables: string, int, array,Array accessing
		//immutable variable, can't be change to other value
		val i = 1
		//mutable variable
		var i1:Int = 1
		val str = "This is a string"
		val myArray:Array[Int] = Array(1,2,3)
		println("The second element of array is: " + myArray(1))
		
		//If…else struct
		if(i == 1)
			println("i is 1")
		else
			println("i is not 1")
			
		//For/looping/while/do while/foreach struct
		//to means: 1,2,3...10
		for(i <- 1 to 10)
			println("i is:" + i)
		//until means 0,1,2
		for(i <- 0 until 3)
			println(myArray(i))
		
		//Writing code in multiple lines
		
		//Comparators: ==, <,>,<=,>=,!=
		
		//Define function/methods
		//Method without return value
		def sayHello(name:String):Unit = println("Hello, " + name)
		sayHello("Bill")
		//a method without parameter and return a string
		def simpleMethod = "Hello"
		//calling a method without parameter shouldn't contain ()
		println(simpleMethod)
		
		//Method with return value
		//if method body contains only one line, 
		//could be: def addOne(number:Int) = number + 1
		def addOne(number: Int) = {
			number+1
		}
		println("4 add one is:" + addOne(4))
		
		//Method, default arguments
		def showNameAgain(firstName:String = "", lastName:String = "") = 
			println(firstName + "," + lastName)
		//if you pass one parameter, this will be passed to first
		showNameAgain("Jary")
		//if you want  to only pass lastName:
		showNameAgain(lastName="Tomlinson")
		
		//Type convert:
		//Int to string
		println(i.toString())
		//String to int
		var a = 1
		println("1+2=" + (a+"2".toInt))
			
		//Get type of object
		println("i=" + i + ", type is " + i.getClass)
		println("str=" + str + ", type is " + str.getClass)
		
		//String:
		//scala string is java.lang.String, so we can use java String function
		//Define a string
		val str1 = "This is a string"
		//Basic string methods:
		//Length
		println(str1 + ", length is:" + str1.length)
		//String match or not
		println(str1 + " match STRING? " + str1.matches("STRING"))
		//Find a substring
		println(str1.substring(3))
		//StartWith:
		println(str1 + " starts with This? " + str1.startsWith("This"))
		//endWith
		println(str1 + " ends with string? " + str1.endsWith("string"))
		//String replace
		println(str1.replace("string","not a string"))
		//String format
		println(String.format("Your name is %s", "Jerry"))
		
		//Regular expression:
		//Define a regular variable
		val pattern = "name".r
		//Basic regular methods:
		//Match or not
		println("123".matches("\\d{3}"))
		//Get subGroups
		
		//Collection:
		//List:
		//Define a list
		val list = List(1,2,3)
		
		//Access item of list
		println(list(0))
		
		//Add an item to the end of list
		println(list + " add 1 to the end: " + (list :+ 1))
		
		//Add an item to beginning of list
		println(list + " add 1 to the beginning: " + (1 +: list))
		
		//Delete an item to list
		//Check an item in list
		println(list.contains(1))
		
		//Set:
		//Define a Set
		val set1 = Set(1,2,3,3)
		println(set1)
		
		//Access item Set
		//Add an item to Set
		//Delete an item to Set
		//Check an item in Set
		println("1 in set?" + set1(1))
		
		//Map:
		//Define a Map
		val map = Map(1->"a",2->"b")
		println(map(2))
		
		//Access item of Map
		println(map.get(1))
		
		//Add an item to Map
		val map1 = map + (3->"c")
		println("old map is:" + map)
		println("new map is:" + map1)
		
		//Delete an item to Map
		//Check an item in Map
		println(map.contains(1))
		
		//Tuple(if have)
		//Define a tuple
		val tp = (1,2)
		//Access item of tuple
		println(tp._1)
		
		//Deconstruct a tuple
		val (first,second) = tp
		println("first item in tuple:" + first)
		println("second item in tuple:" + second)

		println("item type is tuple is:" + first.getClass)
		
		//Class
		//Define class without body
		class Hello
		val t = new Hello
		//print address of t in memory, e.g.Main$Hello@7586beff
		println(t)
		
		//Class with a body, with mutable field(can be changed)
		class Hello1{
			var message :String = "This is a class"
		}
		
		val t1 = new Hello1()
		//print: This is a class
		println(t1.message)
		
		//promoting a parameter to a field
		class Hello2(val message:String){
		}
		
		val t2 = new Hello2("promoted field")
		//print: promoted field
		println(t2.message)
		
		
		//object, it is a singleton class
		object myObj{
			def message = "Hello"
			private val defaulValue = 1
		}
		
		//File I/O:
		//Create file obj
		//Readlines from file
		
		//Language specific features
		//case class: class only contains data
		case class Time(hours:Int=0,minutes:Int=0)
		
		val time = Time(9,0)
		val time1 = Time(9,0)		
		
		//unapply function can retrieve the value inside an case class
		val timeOption = Time.unapply(time)
		//timeOption is Tuple2 wrapped by "Some" type
		println("time case class unapplied is: " + timeOption + "its value is: " + timeOption.get.getClass)
		
		//cass class compared by value, not by reference
		println(time == time1)
		
		//Option:
		var op = Option("Jamie")
		println(op)
		println(op.getClass)
		
		//get value of option
		println(op.get)

		//get value of option, if not found ,return a default value
		println(op.getOrElse("Jane"))
		
		//option can be assigned with Some(option with value) or None(option without value)
		op = None
		println(op)
		op = Some("Jamie")
		println(op)
		
		//Higher order functions take other functions as parameters 
		//or return a function as a result.
		//This is possible because functions are first-class values in Scala. 
		val salaries = Seq(20000, 70000, 40000)
		println("Salary is: " + salaries)
		//return function to multiple a value by 2
		val doubleSalary = (x: Int) => x * 2
		
		//Higher order function often used in map method of List
		val newSalaries = salaries.map(doubleSalary)
		//or using anonymous function
		val newSalaries1 = salaries.map(x=>x * 2)
		println("doubled salary is: " + newSalaries1)
		
		//map function for collection
		//map will return a new map by a function parameter
		val myVector = Vector(1,2,3,4,5)
		var myVector1 = myVector.map(number=>number+1)
		println("Old vector: " + myVector)
		println("New vector: " + myVector1)
		
		//flatMap method for collection
		val fruits = List("Apple", "Pear")
		println("Original map:" + fruits)
		println("Flat map:" + fruits.flatMap(word=>word))
		
		//filter function for collection
		//filter will return a new map, which contains item filtered by a function parameter
		val languages = List("python", "java")
		val python = languages.filter(lang=>lang.contains("py"))
		println("Language contains 'python': " + python)
		
		//foreach function will execute some method for each item in collection
		var listForeach = List(1,2)
		listForeach.foreach(println)

		//forall function(if all elements satisfy condition, return true)
		var result = listForeach.forall(number=>number>100)
		println(result)
		
		//reduce method for collection
		val numbers = 1 to 10
		println("The numbers:" + numbers)
		println("Add them together:" + numbers.reduce(_+_))
		println("Multiple them together:" + numbers.product)
		
		//exist and find(find will only find the first value which matches the condition)
		println("Is there a number which is more than 100?" + numbers.exists(value=>value>100))
		println("Find first value which is less than 7: " + numbers.find(number=>number<=7).get)
		
		//takeWhile(take the item to a new list, until find item which do not match condition)
		//and dropWhile(drop item from list, unit find item which do not matched items)
		//will return an empty list
		println(numbers.takeWhile(num=>num>2))
		//will still return list contains 1 to 10
		println(numbes.dropWhile(num=>num>2))
	}
}

File: python/first.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#the first line is used only in unix when you want to make python file executable
#the second is used to declare a encoding other than utf-8. if no this line, the code file is treated as UTF-8
import re

#Print "hello, world"
print("Hello, world")

#use \ to escapse:
print ('It\'s a desk') #will print It's a desk

#use raw string:
print (r'It\'s a desk') #will print It\'s a desk

#print string in multiple lines
#if string end with "\", new line will not be printed
print ("""abc\
      , it\'s a long string""")
      
#without "\", new line will be printed
print("""this is line one,
        this is line two.""")

#Define variables: string, int, array,Array accessing
myName="this is a string"
myAge=22
print(myName)
print(myAge)

#Array is not standard type in python, you can use list

#If…else if...else struct
#in python, the condition can be with or without brackets
if(myAge > 30):
	print("Your age is older than 30")
elif myAge > 20:
	print("Your age is older than 20")
else:
	print("you are young")
	
#ogical operators(and, or, ==)
print(True and False) #False
print(True or False) #True
print("True or False is: " + str(True or False)) #print True
print("True and False is: " + str(True and False)) #print False
print("True is False?" + str(True == False))#print False

	
#For/looping/while/do while/foreach struct
#for:
words = ['cat','window','defenestrate']
for w in words:
    print (w,len(w))

for num in range(2,10):
    if num %2 == 0:
        print ('Found an even number', num)
    else:
        print ('Found a number',num)
		

#While:
#print a Fibonacci array:
print ("a Fibonacci array")
a,b=0,1 #mutiple assignment
while b<10:
	print (b,",")
	a,b=b,a+b

#flow control in loop: break, continue
for num in range(2,10):
	if num < 5:
		continue
	else:
		print("Find 5 in range 2 to 10")
		break


#Writing code in multiple lines

#Comparators: ==, <,>,<=,>=,!=

#Define function/methods
#Method without return value
def fib(n):
    a,b=0,1
    print ("display a fibonacci array:")
    while a<n:
        print (a,",")
        a,b=b,a+b

#call a function
fib(10)


#Method with return value


#Method, default arguments

#Type convert:
#Int to string
print("This is a num:" + str(1))
#String to int
print(4+int("1"))
#Character to int
print(ord('\n')) #print 10
#int to Character
print(chr(97)) #print lower case 'a'
	
#Get type of object
map={1:'map',2:'ChinaMap'}
print(type(map)) #<class 'dict'>
print(type(map.keys())) #<class 'dict_keys'>
print(type(map.values()))#<class 'dict_keys'>
print(type(1))#<class 'int'>
print(type('abc'))#<class 'str'>
print(type([1,2,3]))#<class 'list'>
print(type((1,2,3)))#<class 'tuple'>
print(type({1,2,3}))#<class 'set'>   

#String:
#Define a 
str1 = "This is a string"

#Basic string methods:
#Length
print (len(str1))

#concatenate two string:
print ("Hello, " + "world!")

#String contains:
print("a" in "abc") #True
print("a" not in "abc") #False

#index a char in string:
print (str1[0]) #print T

#index from the last char:
print (str1[-1])#print g

#index of string
print("Position of 4 in string is: " + str(str1.index('s'))) #return first index of 4, if 4 not exist in string, Excepiton thrown
print("Position of i in string is: " + str(str1.find('i')))#return first index of "i", if "i" not exist in string, return -1


#slicing a string
str1='0123456'

#print string from position 0 (included) to 2(excluded)
print (str1[0:2])

str1 = "    this is a string      "
#String match or not
#Find a substring

#StartWith:

#endWith


#String format
print("Hello, {0}, today is:{1}".format("xiejia","Sunday"))

#String trim:
print(len(str1))
print(len(str1.strip()))
print("String is: " + str1)
print("Trim left    side space: " + str1.lstrip())
print("Trim right side space: " + str1.rstrip())
print("Trim both side space: " + str1.strip())

#String replace
str1= str1.replace("string","orange")
print("Now the string is: " + str1)


#Regular expression:
#Define a regular variable
reg = "\.java:\d+"

#Basic regular methods:
#Match or not
re.search(reg,"RemoteTestRunner.java:382") == None #False

str1="abcdef"
match = re.search("abc",str1)
if(match):
    print(str1 + " match " + "abc")

#match ignore case:
match = re.search("ABC", str1, re.IGNORECASE)
if(match):
    print(str1 + " match " + "ABC(ignore case)")

#Get subGroups
matchs = re.search("\.java:(\d+)","RemoteTestRunner.java:382")
#first group is matched part
matchs.group(0)#.java:382
#second group is part in "()"
matchs.group(1)# 382

#Collection:
#List:
#Define a list
fruits = ["apple","pear","peach"]

#Access item of list
print(fruits[0]) # apple

#Add an item to the end of list
fruits.append("lychee") #['apple', 'pear', 'peach', 'lychee']

#Add an item to beginning of list
fruits.insert(0,"tomato") #['tomato', 'apple', 'pear', 'peach', 'lychee']

#Delete an item from list
fruits.remove("apple") #['tomato', 'pear', 'peach', 'lychee']

#Check an item in list
print("apple" in fruits) #False

#Set:
#Define a Set
book={"book1","book2","book3"}
print(book)


#Access item Set
#Can't access a set by index. using for:
for myBook in book:
    print("I have this book:" + myBook)
    
#Add an item to Set
book.add("Scary book")
print(book)

#remove  item from set. Items should exist in set ,otherwise, exception will be thrown
book.remove("book1")
#remove  item from set no matter item exist in a set
book.discard("book8")

#Check an item in Set
print("book1" in book) #True

#Map:
#Define a Map
myMap = {}

#Add an item to Map
myMap["name"] = "name1"
myMap["value"] = "value1"

#Access item of Map
print("name:" + myMap["name"])
print("value:" + myMap["value"])

#Delete an item to Map
#Check an item in Map
print("name" in myMap.keys()) #print True
print("name1" in myMap.keys()) #print True
print("value1" in myMap.values()) #print True



#Tuple(if have)
#Define a tuple
#Access item of tuple


#Class

#object, it is a singleton class

#File I/O:
#Create file obj
#Readlines from file

#Language specific features

